x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/MlProcessors.java

```diff
@@ -69,6 +69,6 @@ public static Processors getTotalMlNodeProcessors(DiscoveryNodes nodes, Integer
                 total += nodeProcessors.roundDown();
             }
         }
-        return Processors.of((double) total);
+        return total == 0 ? Processors.ZERO : Processors.of((double) total);
```

Comment:
You could use `Processors.Zero` instead of an int for `total` and rely on `Processors#sum`, that should solve the issue.

Reply:
Hmm, that's not really compatible with the "round down before summing"

Reply:
I just looked more closely and in the actual allocation code we're using `roundUp`: https://github.com/elastic/elasticsearch/blob/31ca2f72d3bff000fb2876950391d5b8b1e0a0e8/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentRebalancer.java#L306

So we should use that here too, and that might solve the original problem because presumably no node has zero processors unrounded.

The reason I got the code in this file wrong in the first place was that I copied the logic from here: https://github.com/elastic/elasticsearch/blob/e34ea3be3739283f56a429fb8b87b075686542ef/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java#L85

So that is wrong too - that should also be using `roundUp`.

Sorry for all the confusion. Please could you change all the usages of `roundDown` in the ML code to `roundUp` and adjust the tests to match. That's:

```
src/main/java/org/elasticsearch/xpack/ml/utils/MlProcessors.java:                total += nodeProcessors.roundDown();
src/main/java/org/elasticsearch/xpack/ml/action/TransportMlInfoAction.java:                limits.put("max_single_ml_node_processors", singleNodeProcessors.roundDown());
src/main/java/org/elasticsearch/xpack/ml/action/TransportMlInfoAction.java:                    * singleNodeProcessors.roundDown();
src/main/java/org/elasticsearch/xpack/ml/action/TransportMlInfoAction.java:                limits.put("total_ml_processors", totalMlProcessors.roundDown() + potentialExtraProcessors);
src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java:                .roundDown()
```

Reply:
All fixed.